File: 2024/day-01/day1.py
def part1():
  sum_of_differences = 0
  ls_left = []
  ls_right = []

  with open("./day-01/input.txt", "r") as f:
    for line in f:
      left, right = line.split()
      ls_left.append(int(left))
      ls_right.append(int(right))

    ls_left.sort()
    ls_right.sort()

    for i in range(len(ls_left)):
      difference = abs(ls_left[i] - ls_right[i])
      sum_of_differences += difference

  return sum_of_differences


def part2():
  # Searching each number is O(n) unless we use binary search on a sorted list which is O(logn)
  # Sort and binary search: O(nlogn)
  # Brute force search: O(n^2), could also cache for speedup.
  # Since dataset is relatively small, I'm just going to brute force it.

  ls_left = []
  ls_right = []
  similarityScore = 0

  with open("./input.txt", "r") as f:
    for line in f:
      left, right = line.split()
      ls_left.append(int(left))
      ls_right.append(int(right))
    
  for l in ls_left:
    counter = 0
    for r in ls_right:
      if l == r:
        counter += 1
    similarityScore += l * counter

  return similarityScore

if __name__ == "__main__":
  print(part1())
  print(part2())

File: 2024/day-01/day1.go
package main

import (
  "bufio"
  "fmt"
  "os"
  "sort"
  "strconv"
  "strings"
)

func process_input() ([]int, []int) {
  var lsLeft, lsRight []int

  file, err := os.Open("./day-01/input.txt")
  if err != nil {
    fmt.Errorf("Error opening file: ", err)
    return nil, nil
  }
  defer file.Close()
  
  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    line := strings.TrimSpace(scanner.Text())
    parts := strings.Fields(line)
    left, _ := strconv.Atoi(parts[0])
    right, _ := strconv.Atoi(parts[1])
    lsLeft = append(lsLeft, left)
    lsRight = append(lsRight, right)
  }
  if scanner.Err() != nil {
    fmt.Errorf("Error reading file: ", err) 
  }

  return lsLeft, lsRight
}

func solve_part1(lsLeft, lsRight []int) int {
  var sumOfDifferences int = 0
  sort.Ints(lsLeft)
  sort.Ints(lsRight)

  for i := 0; i < len(lsLeft); i++ {
    var difference int = abs(lsLeft[i] - lsRight[i])
    sumOfDifferences += difference
  }
  return sumOfDifferences
}

func abs(x int) int {
  if x < 0 {
    return -x
  }
  return x
}

func solve_part2(lsLeft, lsRight []int) int {
  var rightCount = make(map[int] int)
  var similarityScore int = 0

  for _, value := range lsRight {
    rightCount[value] = rightCount[value] + 1
  }
  
  for _, value := range lsLeft {
    similarityScore += value * rightCount[value]
  }
  
  return similarityScore
}

func main() {
  var lsLeft, lsRight []int = process_input()
  fmt.Println("Part 1: ", solve_part1(lsLeft, lsRight))
  fmt.Println("Part 2: ", solve_part2(lsLeft, lsRight))
}
